### PR TITLE
Issue #14950: Add new property to UnusedLocalVariable to suppress violtions on unnamed variables

### LINKED ISSUE #14950: Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: UnusedLocalVariable
child of #14942 , #14960

we need to update `UnusedLocalVariableCheck` to support unnamed variables in local variables declaration and in enhanced/basic for.

I have read check documentation: https://checkstyle.org/checks/coding/unusedlocalvariable.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> javac src/Test.java
PS D:\CS\test> cat src/Test.java
import java.util.PriorityQueue;
import java.util.Queue;

class Test {
    void test() {
        Queue<Integer> q = new PriorityQueue<>();
        q.add(1);
        while (!q.isEmpty()) {
            var _ = q.poll();   // violation
        }
    }
}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="UnusedLocalVariable">

    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:13: Unused local variable '_'. [UnusedLocalVariable]
Audit done.
Checkstyle ends with 1 errors.
```
---

**Describe what you expect in detail.**

underscore is used to declare variables that are not intended to be used. So, I consider this example as a false positive.  the check should not violate unnamed variables

---

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #14950: 
Diff Regression config: https://gist.githubusercontent.com/mahfouz72/afd57aa6e51ca161b7927acd959e5211/raw/73fb0891a2bf6813dbebdd94abfb20429efd227b/check.xml

Diff Regression patch config: https://gist.githubusercontent.com/mahfouz72/33967bf86e4a29cafa429771b0e95808/raw/6de640382dd85765496efb44cae1b60a43706115/check_patch.xml

Diff Regression projects: https://gist.githubusercontent.com/mahfouz72/a3d0af030c8f5efd0d8a39f2c14750bc/raw/a3424ad9b6f722e5de1a927e3d85e383101b2b93/projects-to-test-on.properties

### CODE REVIEW

**Comment by romani** [comment]
If you move first commit to separate PR, I will merge it by single review.

**Comment by mahfouz72** [comment]
@romani done https://github.com/checkstyle/checkstyle/pull/15189  please merge ASAP. because this PR will be blocked until the supplemental gets merged
I will drop the supplemental comment here and rebase after merging the other one.

**Comment by mahfouz72** [comment]
github, generate site

**Comment by mahfouz72** [comment]
github, generate report

**Comment by mahfouz72** [comment]
github, generate report

**Comment by mahfouz72** [comment]
github, generate report

**Comment by mahfouz72** [comment]
github, generate report

**Comment by mahfouz72** [comment]
@nrmancuso @rnveach ping

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUnusedLocalVariableTest.java
Comment (rnveach) [inline src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUnusedLocalVariableTest.java:46]: I also disagree here. We should be checking both on and off can be suppressed by xpath.

The point of these tests is to ensure xpath always work. In the beginning, some checks and ASTs could not be suppressed.

**Review by rnveach — CHANGES_REQUESTED** [review]

**Comment by rnveach** [comment]
@mahfouz72 In the future, can you label the regressions for us on what they are if there are multiple and hide the ones we shouldn't be looking at if they are old. I believe GH regression action accepts labels as a parameter but I don't use it so I don't know what it is.

I expect to see regression on
* Old behavior versus new behavior with property turned off. There should be no differences.
  * https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/48f0cc6_2024174138/reports/diff/index.html
* Old behavior versus new behavior with property turned on (this is both default modules). This is where differences should be.
  * https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/48f0cc6_2024161154/reports/diff/index.html

**Comment by rnveach** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/48f0cc6_2024161154/reports/diff/checkstyle/index.html#A28
Am I missing something, why isn't there a removal on line 7? Is our check basically confused because there are 2 variables with the same name?
We ignore try/catch for lines 8-12, correct?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java:58]: > "Allow unnamed variables to be unused.".

sounds weird to me. because unnamed variables can't be used by definition, this is something we can't allow or disallow. 
what we allow or disallow is treating it as a violation or not

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java
Comment (mahfouz72) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java:5]: > I think some wording should also be added that with property on, they can remove the unused variable or convert it to be unnamed.

unused named local variable should be unnamed?

**Comment by mahfouz72** [comment]
> Am I missing something, why isn't there a removal on line 7? Is our check basically confused because there are 2 variables with the same name?

There was no violation before check got confused on [line 8](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/48f0cc6_2024161154/reports/diff/checkstyle/xref/home/runner/work/checkstyle/checkstyle/.ci-temp/contribution/checkstyle-tester/repositories/checkstyle/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/InputUnnamedVariableBasic.java.html#L8).  and registered `_` as used because it is encountered in resource specification in try 

```
PS D:\CS\test> cat src/Test.java                                                                                               
public class Test {
    void m() {
        int _ = 0; // ok
        try (final Lock _ = null) { } // check recognized '_' in resource as usage of the above variable
    }
    void m2() {
        int _ = 0; // violation
    }
    class Lock implements AutoCloseable {
        @Override
        public void close() {}
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:7:9: Unused local variable '_'. [UnusedLocalVariable]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```
---
> We ignore try/catch for lines 8-12, correct? 

Yes, the check currently ignores try we should add support for try in this check in another issue and the exception parameter is not even considered a local variable so it is ik to get ignored by this check, catch parameters needs a new check

---
> I am reading this right and this file isn't compilable and so not an issue with this check?

Yes

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java
Comment (mahfouz72) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java:5]: > Also, add some new unnamed examples in this file

This file has no space left for any extra examples. should I try to remove some examples in this file to have a space for unnamed examples? Or continue in a new file but then we duplicate the config again or just not showing it?

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/InputUnusedLocalVariableWithAllowUnnamedFalse.java
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/InputUnusedLocalVariableWithAllowUnnamedFalse.java]: done

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUnusedLocalVariableTest.java
Comment (mahfouz72) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUnusedLocalVariableTest.java]: done

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java
Comment (mahfouz72) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java]: > Also, add some new unnamed examples in this file and then introduce a 2nd example with the property turned off. Use the same Java code between them so its clear this is what changed when changing the property.

done but has to made some minor changes in the file to save space for new unnamed examples

**Comment by mahfouz72** [comment]
github, generate site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java:58]: > unnamed variables can't be used by definition

I was thinking more along the lines of also old Java versions where `_` could be assigned. Do we need to be more specific and says "single underscore variables"? @nrmancuso What you think?

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java
Comment (rnveach) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java]: > to save space 

Why do you need to save space?

Comment (mahfouz72) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java]: CS config, file length limit is 60.

**Comment by rnveach** [comment]
> There was no violation before check got confused on [line 8](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/48f0cc6_2024161154/reports/diff/checkstyle/xref/home/runner/work/checkstyle/checkstyle/.ci-temp/contribution/checkstyle-tester/repositories/checkstyle/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/InputUnnamedVariableBasic.java.html#L8). and registered _ as used because it is encountered in resource specification in try

So if we had `int _ = 0;` and `int _ = 1;` alone in a method, it would be 2 violations? It was the try-with-resource which was throwing it off?

> registered _ as used because it is encountered in resource specification in try

Isn't this basically a bug? We are declaring a variable here, it is not usage.

Also, did old Java versions never complain about a duplicate variable for `_`? 3 things in this example is named the same, but there is no compile error?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java:58]: yes IMO if we want to be accurate on this we should use "underscore" not "unnamed"

File: src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example2.java
Comment (rnveach) [inline src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example2.java:34]: Where is the variable `i`?

**Review by rnveach — CHANGES_REQUESTED** [review]

**Comment by mahfouz72** [comment]
> So if we had int _ = 0; and int _ = 1; alone in a method, it would be 2 violations? It was the try-with-resource which was throwing it off?

Yes, these 2 will be a violation if there is no try with resource in the file.

> Isn't this basically a bug? We are declaring a variable here, it is not usage.

This will be a bug for people who have used `_` as a normal identifier before (java 9) if they declare some local variable named `_` and a resource `_` in a try then the resource will cancel the violation of the local variable even if the variable is not used. for people who used `_` as unnamed it will not be a bug because we can't use this variable so it can't be violated as unused

---
> Also, did old Java versions never complain about a duplicate variable for _? 3 things in this example is named the same, but there is no compile error?

I am not sure if I understand this but yes it is normal to compile we can have more than one unnamed variable in a file (after Java 21), before java 9 this wouldn't compile because it was a valid identifier

File: src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example2.java
Comment (mahfouz72) [reply src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example2.java:34]: sorry copy-paste by mistake

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java:58]: `AllowUnnamedVariables` - do not place violation on variables named with a single underscore (known as `unnamed variables in Java 21+).

Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java]: done

**Comment by mahfouz72** [comment]
github, generate site

File: src/xdocs/checks/coding/unusedlocalvariable.xml.template
Comment (nrmancuso) [inline src/xdocs/checks/coding/unusedlocalvariable.xml.template:47]: Please stick to language like `Do not place violations on variables named with a single underscore (known as <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
 unnamed variables</a> in Java 21+).`, `underscore variables` is nonsense 😂

**Review by nrmancuso — CHANGES_REQUESTED** [review]
I am good after https://github.com/checkstyle/checkstyle/pull/15187/files#r1670707087

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java
Comment (rnveach) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/Example1.java]: Ok, I forgot we set that limit on examples. That would seem to be an issue for these type of examples, meaning we would have to split these type up more in the future.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java]: @nrmancuso I am un-resolving this. Property documentation need to start with a verb, otherwise the generated javadocs don't make sense.

> Do not place violations on variables ...
> Setter to do not place violations ...

**Review by rnveach — APPROVED** [review]
I am good, but the following need to be done:

https://github.com/checkstyle/checkstyle/pull/15187#discussion_r1670929685
https://github.com/checkstyle/checkstyle/pull/15187#issuecomment-2218293337

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java]: > @nrmancuso I am un-resolving this. Property documentation need to start with a verb, otherwise the generated javadocs don't make sense.
> 
> > Do not place violations on variables ...
> > Setter to do not place violations ...

Ah shoot. Yes, this is nonsense. 

`Allow variables named with a single underscore (known as unnamed variables in Java 21+). `

Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java]: done

File: src/xdocs/checks/coding/unusedlocalvariable.xml.template
Comment (mahfouz72) [reply src/xdocs/checks/coding/unusedlocalvariable.xml.template]: done

**Comment by mahfouz72** [comment]
>if they declare some local variable named _ and a resource _ in a try then the resource will cancel the violation of the local variable

@rnveach  sorry I missed something. this won't compile before Java 9. You can't have two variables with the same name. accordingly, This is not a bug that affects either Java 9 or Java21+.  Do you agree?

**Review by nrmancuso — APPROVED** [review]

**Comment by nrmancuso** [comment]
> I am good, but the following need to be done:
> 
> [#15187 (comment)](https://github.com/checkstyle/checkstyle/pull/15187#discussion_r1670929685) [#15187 (comment)](https://github.com/checkstyle/checkstyle/pull/15187#issuecomment-2218293337)

Based on this comment, I will let @rnveach merge this PR once he is good.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java]: Looks good.

**Comment by rnveach** [comment]
@mahfouz72 Isn't this a bug before Java 9 that there aren't 2 violations, both on the `_` variables and not the resource?

````
$ cat TestClass.java
public class TestClass {
    void method1() {
int _ = 0;
    }

    void method2() {
int _ = 0;

try (Lock _ = null) {}
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <property name="haltOnException" value="false"/>

    <module name="TreeWalker">
<module name="UnusedLocalVariable"/>
    </module>
</module>

$ java -jar checkstyle-10.17.0-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:3:1: Unused local variable '_'. [UnusedLocalVariable]
Audit done.
````

**Comment by rnveach** [comment]
ok, maybe I will look into a bit but it seems a concern that this check thinks a try-with-resource is a usage. My concerns are resolved for now.

**Comment by mahfouz72** [comment]
> maybe I will look into a bit but it seems a concern that this check thinks a try-with-resource is a usage

It is a concern of course but doesn't result in a bug. if it works don't touch it : ) 

> My concerns are resolved for now.

Then please merge when you can this is left for you to merge

**Comment by rnveach** [comment]
> if it works don't touch it 

You can't prove it "works" if we can't say it will never be a problem. The only way we have seen it is with duplicate variable name which isn't valid. We haven't ruled out all circumstances, which is what I want to try and look into.

BTW, isn't this still a bug on Java 21+ if you don't suppress unnamed variables? Nothing says this has to be enabled if you are on a high enough java version.

**Comment by mahfouz72** [comment]
> isn't this still a bug on Java 21+ if you don't suppress unnamed variables?

yes I think, but it is very nonsense if I am on java 21+ and want to violate all unnamed variables as unused

**Comment by mahfouz72** [comment]
If you want I can opem issue to remember this once I am on laptop

**Comment by rnveach** [comment]
Continuing on for bug at https://github.com/checkstyle/checkstyle/pull/15187#issuecomment-2218293337,

It appears [our `IDENT` handling ](https://github.com/checkstyle/checkstyle/blob/947002a3d66557d7e62a5aee81c7de449ce5f168/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java#L245)might be too aggressive. [If it finds an ident which doesn't match a few exclusions we register it as a possible variable usage.](https://github.com/checkstyle/checkstyle/blob/947002a3d66557d7e62a5aee81c7de449ce5f168/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java#L314-L317)

It picked up `Lock _` because `_` is an ident and it looks to see if it is a variable and if so, it is being "used".

I am going to run regression with a change and see if anything pops up.

**Comment by rnveach** [comment]
@mahfouz72 This looks like a valid case for the bug report, right?

https://rveach.no-ip.org/checkstyle/regression/1/openjdk17/index.html#A3
Change: https://github.com/checkstyle/checkstyle/compare/master...rnveach:checkstyle:pr_15187_test

**Comment by rnveach** [comment]
https://rveach.no-ip.org/checkstyle/regression/1/openjdk17/index.html#A3
It looks like we can't simply ignore resources altogether. This is a usage.

**Comment by mahfouz72** [comment]
> It looks like we can't simply ignore resources altogether.

yeah, then it should be ignored onlt if it is a new variable declaration not a usage of previous variable

**Comment by rnveach** [comment]
https://rveach.no-ip.org/checkstyle/regression/2/
https://github.com/checkstyle/checkstyle/compare/master...rnveach:checkstyle:pr_15187_test

Second regression showed less cases.

This still seems like a concern that `ident` is being so far reaching. I am not sure if we could satisfy pitest to merge a PR even if we did make an issue.

I am going to merge this PR for now.

**Pull request merged**
